ethcore/src/tx_filter.rs

```diff
@@ -0,0 +1,233 @@
+	pub const BASIC: u32 = 0b00000001;
+	pub const CALL: u32 = 0b00000010;
+	pub const CREATE: u32 = 0b00000100;
+	pub const _PRIVATE: u32 = 0b00001000;
+}
+
+/// Connection filter that uses a contract to manage permissions.
+pub struct TransactionFilter {
+	contract: Mutex<Option<Contract>>,
+	client: RwLock<Option<Weak<BlockChainClient>>>,
+	contract_address: Address,
+	permission_cache: Mutex<HashMap<(H256, Address), u32>>,
+}
+
+impl TransactionFilter {
+	/// Create a new instance if address is specified in params.
+	pub fn from_params(params: &CommonParams) -> Option<TransactionFilter> {
+		params.transaction_permission_contract.map(|address|
+			TransactionFilter {
+				contract: Mutex::new(None),
+				client: RwLock::new(None),
+				contract_address: address,
+				permission_cache: Mutex::new(HashMap::new()),
+			}
+		)
+	}
+
+	/// Clear cached permissions.
+	pub fn clear_cache(&self) {
+		self.permission_cache.lock().clear();
+	}
+
+	/// Set client reference to be used for contract call.
+	pub fn register_client(&self, client: Weak<BlockChainClient>) {
+		*self.client.write() = Some(client);
+	}
+
+	/// Check if transaction is allowed at given block.
+	pub fn transaction_allowed(&self, parent_hash: &H256, transaction: &SignedTransaction) -> bool {
+		self.client.read().as_ref().map_or(false, |client| {
+			let mut cache = self.permission_cache.lock(); let len = cache.len();
+			let client = match client.upgrade() {
+				Some(client) => client,
+				_ => return false,
+			};
+			let tx_type = match transaction.action {
+				Action::Create => tx_permissions::CREATE,
+				Action::Call(address) => if client.code_hash(&address, BlockId::Hash(*parent_hash)).map_or(false, |c| c != KECCAK_EMPTY) {
+					tx_permissions::CALL
+				} else {
+					tx_permissions::BASIC
+				}
+			};
+			let sender = transaction.sender();
+			match cache.entry((*parent_hash, sender)) {
+				Entry::Occupied(entry) => *entry.get() & tx_type != 0,
+				Entry::Vacant(entry) => {
+					let mut contract = self.contract.lock();
+					if contract.is_none() {
+						*contract = Some(Contract::new(self.contract_address));
+					}
+
+					let permissions = match &*contract {
+						&Some(ref contract) => {
+							contract.allowed_tx_types(
+								|addr, data| futures::done(client.call_contract(BlockId::Hash(*parent_hash), addr, data)),
+								sender,
+							).wait().unwrap_or_else(|e| {
+								debug!("Error callling tx permissions contract: {:?}", e);
+								tx_permissions::NONE
+							})
+						}
+						_ => tx_permissions::NONE,
+					};
+
+					if len < MAX_CACHE_SIZE {
+						entry.insert(permissions);
+					}
+					trace!("Permissions required: {}, got: {}", tx_type, permissions);
+					permissions & tx_type != 0
+				}
+			}
+		})
+	}
+}
+
+impl ChainNotify for TransactionFilter {
+	fn new_blocks(&self, imported: Vec<H256>, _invalid: Vec<H256>, _enacted: Vec<H256>, _retracted: Vec<H256>, _sealed: Vec<H256>, _proposed: Vec<Bytes>, _duration: u64) {
```

Comment:
now that we're adding so many new chain-notify implementations, would be nice to not be cloning each of these vectors each time, especially since most of them only read.

Reply:
Agreed, although most of the time they only contain 0 or 1 elements.